rllib/algorithms/sac/sac.py

```diff
@@ -371,3 +435,149 @@ def get_default_policy_class(
+                        lambda w: w.sample(),
+                        local_worker=False,
+                    )
+                # Perform SAC postprocessing (Prio weights) on a (flattened)
+                # list of Episodes.
+                # TODO (simon): Change this to a preprocessing in the learner.
+                # postprocessed_episodes = self.postprocess_episodes(episodes)
+
+            episodes = tree.flatten(episodes)
+            # TODO (sven): single- vs multi-agent.
+            self._counters[NUM_AGENT_STEPS_SAMPLED] += sum(len(e) for e in episodes)
+            self._counters[NUM_ENV_STEPS_SAMPLED] += sum(len(e) for e in episodes)
+
+            # Add the sampled experiences to the replay buffer.
+            self.local_replay_buffer.add(episodes)
+
+        # Update the target network each `target_network_update_freq` steps.
+        current_ts = self._counters[
+            NUM_AGENT_STEPS_SAMPLED
+            if self.config.count_steps_by == "agent_steps"
+            else NUM_ENV_STEPS_SAMPLED
+        ]
+
+        # If enough experiences have been sampled start training.
+        if current_ts >= self.config.num_steps_sampled_before_learning_starts:
+            # Run multiple training iterations.
+            for _ in range(sample_and_train_weight):
+                # Sample training batch from replay_buffer.
+                train_dict = self.local_replay_buffer.sample(
+                    num_items=self.config.train_batch_size,
+                    n_step=self.config.n_step,
+                    gamma=self.config.gamma,
+                )
+
+                # TODO (sven): This looks a bit ugly, but we have not the same
+                # naming in `SingleAgentEpisode` and `SampleBatch`.
+                # Note, the `SingleAgentEpisode` stores obs_t. action_(t-1),
+                # reward_t, infos_t, while the replay buffer's `sample()`
+                # returns obs_t, action_t, reward_(t-1). So, we take the
+                # reward from the subsequent experience in the episode to
+                # calculate the Q-values.
+                train_batch = SampleBatch(train_dict)
+                # train_batch = SampleBatch(
+                #     {
+                #         SampleBatch.OBS: train_dict[SampleBatch.OBS],
+                #         SampleBatch.NEXT_OBS: train_dict[SampleBatch.NEXT_OBS],
+                #         SampleBatch.ACTIONS: train_dict[SampleBatch.ACTIONS],
+                #         SampleBatch.REWARDS: train_dict[SampleBatch.REWARDS],
+                #         SampleBatch.TERMINATEDS: train_dict["is_terminated"],
+                #         SampleBatch.TRUNCATEDS: train_dict["is_truncated"],
+                #     }
+                # )
+                # Convert to multi-agent batch as `LearnerGroup` depends on it.
+                train_batch = train_batch.as_multi_agent()
+
+                def reduce_fn(results: List[ResultDict]) -> ResultDict:
```

Comment:
Let's add a TODO here to streamline these custom metrics reduction functions via the Learner's register_metric(s) APIs.